### DISCUSSION TITLE
Any plan for input stream manipulation support?

### DISCUSSION THREAD

**Discussion by user202729** [lvjr/functional#14] [General]
As in the title. I'm taking about functions such as...

```
\futureLetToNextToken \lTmpaToken  % assign tmpa to the meaning of the next token
\futureLetToNextTwoToken \lTmpaToken  % assign tmpb to the meaning of the next-next token
\peekAnalyzeNextToken \lTmpaTl \lCharcodeTl \lCatcodeTl  % assign these 3 variables to the token, catcode, char code etc.
\peekAnalyzeReturnNextToken \lTmpaTl  % return the token
\grabDelimitedArgument \lTmpaTl {#1 abc}
\grabXparseArgument \lTmpaTl {m}  % grab an argument and store it to \lTmpaTl
\grabXparseArgument \lTmpaTl {o}
```

Currently these are not implement-able.

Side note, they would be pretty slow, as each time they are called they need to skip through all the "pending code" following in the input stream. (alternative includes making a "compiler" that compiles the code to "normal" code, that way you don't have the disadvantage of another layer of interpreter. I'm doing something similar with one of my packages)

**Comment by lvjr** [comment]
It is quite complicated to add them. Maybe I would not make `functional` package go that far in the near future.

**Reply by user202729** [reply]
It's not that complicated, just slow. You need to
• keep a unique marker at the end of the top layer of the input stream
• for such functions, temporary store the content until the unique marker somewhere, do the necessary work, then restore the input stream
(some implementation note, make sure no brace group is removed, and make sure there's no explicit brace character imbalance)

if you want to implement `return` you need some unique marker anyway (although it's at each layer)
